src/System.Collections.Sequences/System/Collections/Sequences/Position.cs

```diff
@@ -1,47 +1,64 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using System.Collections.Generic;
+using System.ComponentModel;
 using System.Runtime.CompilerServices;
 
 namespace System.Collections.Sequences
 {
     public struct Position : IEquatable<Position>
     {
-        public object ObjectPosition;
-        public int IntegerPosition;
-        public int Tag;
+        long _index;
+        object _obj;
 
-        public static readonly Position First = new Position();
-        public static readonly Position AfterLast = new Position() { IntegerPosition = int.MaxValue - 1 };
+        public static explicit operator int(Position position) => (int)position._index;
+        public static implicit operator long(Position position) => position._index;
+        public T As<T>() => _obj==null||IsInfinity?default:(T)_obj;
+        public bool IsInfinity => this == Infinity;
+        public bool IsDefault => _obj == null && _index == 0;
 
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static bool operator==(Position left, Position right)
-        {
-            return left.Equals(right);
-        }
+        public static readonly Position Infinity = new Position(new object(), long.MaxValue);
 
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static bool operator!=(Position left, Position right)
-        {
-            return !left.Equals(right);
-        }
+        public static Position operator +(Position position, int value) => new Position(position._obj, position._index + value);
+        public static Position operator +(Position position, long value) => new Position(position._obj, position._index + value);
+
+        public static Position operator -(Position position, int value) => new Position(position._obj, position._index - value);
+        public static Position operator -(Position position, long value) => new Position(position._obj, position._index - value);
+
+        public static bool operator ==(Position left, Position right) => left._index == right._index && left._obj == right._obj;
+        public static bool operator !=(Position left, Position right) => left._index != right._index || left._obj != right._obj;
+
+        public void Set(int value) => _index = value;
```

Comment:
`int` is implicitly convertible to `long`, so there are no reasons to have `Set(int)`.

Reply:
Will remove.

Reply:
Actually, I do need the int overload. Otherwise Set<T> is called on int argument.

Reply:
Maybe `T` should be limited by the `class` constraint? The value is stored as `object` so that limitation will prevent implicit boxing inside `Set<T>`. As a bonus - performance optimization because `EqualityComparer<T>.Default.Equals(value, default)` will be replaced by `value == null`.

Reply:
Yeah, I like the idea. Thanks.